pkg/gofr/datasource/file/ftp/file_test.go

```diff
@@ -0,0 +1,632 @@
+
+			err := reader.Scan(content)
+
+			assert.Error(t, err)
+			assert.Equal(t, "", content)
+		}
+	})
+}
+
+func Test_ReadFromJSONArray(t *testing.T) {
+	runFtpTest(t, func(fs *ftpFileSystem) {
+		var jsonContent = `[{"name": "Sam", "age": 123},{"name": "Jane", "age": 456},{"name": "John", "age": 789}]`
+
+		type User struct {
+			Name string `json:"name"`
+			Age  int    `json:"age"`
+		}
+
+		var jsonValue = []User{{"Sam", 123}, {"Jane", 456}, {"John", 789}}
+
+		newCsvFile, _ := fs.Create("temp.json")
+
+		_, _ = newCsvFile.Write([]byte(jsonContent))
+		newCsvFile, _ = fs.Open("temp.json")
+
+		reader, _ := newCsvFile.ReadAll()
+
+		defer func(fs datasource.FileSystem, name string) {
+			_ = fs.RemoveAll(name)
+		}(fs, "temp.json")
+
+		var i = 0
+
+		for reader.Next() {
+			var u User
+
+			err := reader.Scan(&u)
+
+			assert.Equal(t, jsonValue[i].Name, u.Name)
+			assert.Equal(t, jsonValue[i].Age, u.Age)
+
+			assert.NoError(t, err)
+
+			i++
+		}
+	})
+}
+
+func Test_ReadFromJSONObject(t *testing.T) {
+	runFtpTest(t, func(fs *ftpFileSystem) {
+		var jsonContent = `{"name": "Sam", "age": 123}`
+
+		type User struct {
+			Name string `json:"name"`
+			Age  int    `json:"age"`
+		}
+
+		newCsvFile, _ := fs.Create("temp.json")
+
+		_, _ = newCsvFile.Write([]byte(jsonContent))
+		newCsvFile, _ = fs.Open("temp.json")
+
+		reader, _ := newCsvFile.ReadAll()
+
+		defer func(fs datasource.FileSystem, name string) {
+			_ = fs.RemoveAll(name)
+		}(fs, "temp.json")
+
+		for reader.Next() {
+			var u User
+
+			err := reader.Scan(&u)
+
+			assert.Equal(t, "Sam", u.Name)
+			assert.Equal(t, 123, u.Age)
+
+			assert.NoError(t, err)
+		}
+	})
+}
+
+func Test_ReadFromJSONArrayInvalidDelimiter(t *testing.T) {
+	runFtpTest(t, func(fs *ftpFileSystem) {
+		var jsonContent = `!@#$%^&*`
+
+		newCsvFile, _ := fs.Create("temp.json")
+
+		_, _ = newCsvFile.Write([]byte(jsonContent))
+
+		newCsvFile.Close()
+
+		newCsvFile, _ = fs.Open("temp.json")
+
+		_, err := newCsvFile.ReadAll()
+
+		defer func(fileStore datasource.FileSystem, name string) {
+			_ = fileStore.RemoveAll(name)
+		}(fs, "temp.json")
+
+		assert.IsType(t, &json.SyntaxError{}, err)
+	})
+}
+
+// Helper function to run FTP tests requiring actual ftp connection.
+// LoadConfig loads FTP configuration from environment variables.
+func LoadConfig() *Config {
+	err := godotenv.Load("./configs/.env")
+	if err != nil {
+		log.Printf("Error loading .env file: %v", err)
+	}
+
+	config := Config{
+		Host:      os.Getenv("FtpHost"),
+		User:      os.Getenv("FtpUser"),
+		Password:  os.Getenv("FtpPassword"),
+		Port:      os.Getenv("FtpPort"),
+		RemoteDir: os.Getenv("FtpRemoteDir"),
+	}
+
+	return &config
+}
+
+func runFtpTest(t *testing.T, testFunc func(fs *ftpFileSystem)) {
+	t.Helper()
+
+	config := LoadConfig()
+
+	ftpClient := New(config)
+
+	val, ok := ftpClient.(*ftpFileSystem)
+	if ok {
+		logger := logging.NewMockLogger(logging.DEBUG)
+
+		val.UseLogger(logger)
+		val.Connect()
+	}
```

Comment:
I doubt if it needs to be checked if ftpFileSystem needs to be implemented

Reply:
Connect is a method of filesystem provider interface that embeds the filesystem interface created....New returns a filesystem interface, so I convert it into its concrete type and then call the method connect.